### ISSUE TITLE
[Feature Request] Seperate snatched and downloaded episodes

### ISSUE THREAD

**Issue by selthius** [SiCKRAGETV/sickrage-issues#674]
Currently, there is no way to see which episodes in what quality you have when sickrage has snatched new episodes. A seperation of the statuses is going to make this possible. It will also make it easier to revert to the previous quality if the download is bad.

**Comment by TehWardy** [comment]
Surely ...
What you have is typically in your list on the main page (which shows counts).
What you have snatched is in the backlog.

I tend to assume if it's not in my backlog then I have it since there is no "downloaded" state for backlog items (which makes sense to me).

I tend to pull files, if they are bad i delete and then "rescan" the show which adds the now deleted item back on the backlog. 
Either I'm happy with the one I have or I delete it.

Suggestion:
Maybe having a flag to say "this is here but look for a better copy" might be useful the results of this extra fetching should probably be managed on a new "duplicates" page?

Thoughts?

**Comment by selthius** [comment]
I rarely check my backlog, if ever, because there never seems to be any need for it. I am happy with the information I get from the main view, with the exception of this. The separation would "fix" (in my opinion) two things. Firstly, it would be able to easily revert back to showing the downloaded quality, which is not that easy to do manually if you have a lot of shows and doesn't include the quality in the filename. Secondly, you wouldn't have to experience empty shows in your main view when the first and only season of a show has been snatched in a better quality. There is probably a good way to differentiate snatched and downloaded too in that main view, but that is besides the point.

I do understand that this may/will make sickrage incompatible with sickbeard's database when changing system, but in my opinion this is a much better way to organize episodes, and users will benefit greatly from this. That is also very dependent on how they use sickrage, but everyone who has multiple shows they want in the best quality possible, will probably be very happy for this change.

**Comment by TehWardy** [comment]
I don't understand what you're saying here? The downloaded quality I my case is the quality I have in my library ... Either that's good enough or it isn't which is why I could see a use for a flag to say "this isn't" but don't get why you would intentionally Download new copies of files without a reason ... Imo if sr is aware that you have rejected a file it can fix the problem but you're basically asking for a very manual system where you want to compare 2 files yourself defeating the whole purpose of sb / sr which is to automate downloading.

I personally have a larger library than most (roughly 10TB) and can see no need for any change to the homepage along these lines.

**Comment by selthius** [comment]
Okey, I will try to explain this once more: I want to seperate the snatched and downloaded status, so that when a download fails, because the torrent has been abandoned, the downloaded quality still exists and was not overwritten when the new quality was snatched. This makes it possibly to easily revert to the previous quality. In addition, this would make it so that shows with only 1 season does not appear empty when the whole season has been snatched in a better quality.

I'll give an example just for clarity:

I have currently all the episodes of "The Knick" downloaded in HDTV. Eventually, this show will be released on bluray. My sickrage is configured to always have the best quality available, so when the bluray-release is available, it will be snatched. When this happens, the main view of sickrage will whow that "The Knick" has no episodes downloaded. Only when i click the show to see the details, I will know that it's snatched. Let's say, for arguments sake, that the bluray-release didn't complete, because it is already abandoned. So, now, "The Knick" would still appear empty in sickrage, and never be completed, unless i fix it manually or a better quality is released.

With my proposition, I could just have pressed a button or something, and the quality would be back to hdtv, which is the quality of the files I currently have. It would also, if implemented, be able to inform me that the season is already downloaded, but also snatched in a better quality.

**Comment by TehWardy** [comment]
Ah ok now I get it ...

If I understand you correctly then: 
So you're saying - If you already have HDTV copies of the episodes in question then the homepage should reflect that even if you have snatched new versions at bluray quality.

I completely agree with that, the homepage should show what I have at all times if I both have a file and have a file snatched then it may be useful to have a means to show separately what I have and what I have snatched as currently a snatched file reports as only snatched on the homepage and not downloaded.

Maybe some form of "re-snatched" stat or something could help.

Triggering a re-snatch however is likely something that would be done from either the show page or the bulk status management.

**Comment by selthius** [comment]
Yes, exactly. Your solution is not ideal though, i feel. This either makes it impossible to see the quality you have while it downloads, or you need to have probably 100 different strings to match all the previous qualities with new quality (e.g. hdtv -> 1080p hdtv and 720 web -> 1080p web). I think my solution is better, but may be wrong, so please correct me if I am. This separation of the snacthed and downloaded variables would probably make our database incompatible with sickbeard, and vice versa, but hopefully they will also notice the benefit of this change and do the same.

**Comment by TehWardy** [comment]
Well correct me if I'm wrong but basically you should only ever have at most 2 copies of a file to worry about ....

The copy already in your library.
The snatched copy.

When the snatched copy is downloaded post processing will decide which is kept.

So at most on the homepage you have as it currently is 3 main stats:

Downloaded 
Snatched 
Then possibly this new "re-snatched" figure.

Or am I missing something?

**Comment by selthius** [comment]
The problem is that currently there is no way of knowing the quality of the file in you library, when the new one is snatched, at least to my knowledge. If there is, this should be much more visible. So, I want to make snatched a separate stat, so that we end up with 2, the one already downloaded, and the one which is snatched. As I'm typing this I realize that this must already be the case, because I'm really stupid and can't think things through. So, the issue is just that we need to display the snatched and downloaded stat separately, at least when someone wants to (I have yet to find a option for this).

**Comment by TehWardy** [comment]
Lol finally the penny drops! 
Sounds like we now agree! 
The missing info in this case is in the show view (as I'm sure you figured out).

Time for @fernandog to take over :)

**Comment by fernandog** [comment]
@TehWardy can you summarize text all this please?! =)

**Comment by fernandog** [comment]
@selthius "The problem is that currently there is no way of knowing the quality of the file in you library"
 yes you can. Just open the showpage. there's a column "quality"

**Comment by TagForce** [comment]
@fernandog I think that's the quality of the snatched episode. It overwrites the episode status in the db with snatched/quality, I think.

**Comment by fernandog** [comment]
@TagForce I see. so when the snatch fails it should get the previous status? Another field in the DB? 

Or force a rescan in folder return the previous status?

**Comment by TehWardy** [comment]
When doing a scan get the quality of the file from the file.

When snatching get quality as you currently do.

Best quality wins!

**Comment by fernandog** [comment]
not the best quality wins. if the snatched turn out to be a bad nzb or low seed torrent, it will cancel the snatch status and rescan go get previous status. Add mark that NZB or torrent release as bad and try another

@TagForce agree? you PR about your snatched for x days PR

**Comment by selthius** [comment]
It would be nice if there is possible to see the quality of the previously downloaded file, even when a better version is snatched. If the torrent is a whole season, it can take a while to complete, and it would be nice to see which quality i currently have and am getting, so I can use that information when I decide what to watch next. If it's difficult to fit the information in the design, then maybe some functionality to "disregard" the snatch and show the downloaded file again, the snatched episode will be postprocessed eventually anyway.

**Comment by fernandog** [comment]
@selthius you don't you check HISTORY tab?! showpage don't show history

**Comment by TagForce** [comment]
@TehWardy @fernandog there's no reliable way to get the quality from a postprocessed file, otherthan actually reading it. Snatched status should be seperated from episode status in the db. That's the only failsafe way to keep the quality of the old episode.

**Comment by selthius** [comment]
I do read the history, but if the file was postprocessed ages ago, there is no way to know the quality, without either having it in the filename or have the memory of an elephant. Like @TagForce is saying, the only failsafe way is to have both the previous quality and the snatched quality in the databse, which it probably already is, but it would be great if it could be displayed somewhere too.

**Comment by fernandog** [comment]
It's not. need to upgrade database

**Comment by TehWardy** [comment]
@TagForce @fernandog  @selthius - Ok guys this is how I see it ...

Homepage currently shows per show:
    downloaded + snatched / total.
    show quality to snatch

EXISTING FUNCTIONALITY: 
If a show file is deemed downloaded but not ideal in quality it can be marked as failed / wanted to be re-requested. 

Post process is also smart enough to handle a situation where it has 2 files and pick the best one.


NEW FUNCTIONALITY:
SR does not reflect the actual state correctly on the homepage if I was to mark a file wanted and it already exists in my downloaded library.
If we mark an existing downloaded file as say "wanted" then the homepage will reduce the episode count we have by 1 which is not true.

When we mark a show as "wanted" and it already exists, SR should not reduce the total eps we already have (as we still have it) it should also some how show that although we have this episode as part of the subset that we still need (so its effectively both downloaded and snatched).

The show page could perhaps show in more detail that the ep is both downloaded and snatched and at what quality.


OTHER:
Say I have all the eps for a show and I change the quality up then do a rescan of that show, SR should backlog the whole show as the quality is not high enough (I believe this is existing functionality that even SB does).

It may also be good to be able to say to SR "I want to set this ep as wanted at a specific quality"

----

@selthius 
Is my summary here correct enough that @TagForce  and @fernandog  can roll with this and devise a suitable solution?

**Comment by selthius** [comment]
Yeah, that looks great!

**Comment by tdev5** [comment]
@selthius, thanks for tagging my issue.

All, here is a mockup of an idea for the display. It might also be useful for thinking through the database schema.

Notice:
- Status, downloaded, and snatched become separate columns.
- New "upgrading" status.
- Still one color per row (status).
- "Upgrading" episodes are included in the "low quality" count.

![status_snatched_downloaded](https://cloud.githubusercontent.com/assets/12114482/7338596/f532591c-ec3f-11e4-8ab3-8f401c539a22.png)

**Comment by tdev5** [comment]
Implementing these changes in the database would probably support #1420.

**Comment by fernandog** [comment]
This need a major change in SR and SR database

**Comment by TehWardy** [comment]
Uh oh!

**Comment by tdev5** [comment]
> This need a major change in SR and SR database

Do the database tables that the episode statuses come from look kind of like what's displayed on the show pages? *If so*, it seems like changing it to something like what I pictured above would mainly be adding a few columns. I'm sure there would be lots of little tasks surrounding it, but...what am I missing?

**Comment by TehWardy** [comment]
Am I missing something here?
Where we already have a row in the library we are saying that the state of the row is the result of a compare against what we really want and what we have so ...
These rows are effectively in multiple states ... 1 downloaded state, 1 snatched state.

A colour on a row only portrays 1 state. 
So I would use colour as minimally as possible, perhaps just have neutral traffic light style colouring (quality is, poor / good), and maybe use amber for "poor but snatched a good quality version".

I can't help but feel that you missed something but can't put my finger on it.

**Comment by tdev5** [comment]
OK, I have some ideas to simplify... Details later.

**Comment by tdev5** [comment]
>  Where we already have a row in the library we are saying that the state of the row is the result of a compare against what we really want and what we have so ...

I don't understand what you're saying here.

> These rows are effectively in multiple states ... 1 downloaded state, 1 snatched state.

That's one way of looking at it. And I think the downloaded state is more important. It says what's actually in the library, instead of what might be later.

So how about this...the "upgrading" status shows as the same color as "downloaded".

![status_snatched_downloaded](https://cloud.githubusercontent.com/assets/12114482/7362304/7740b162-ed56-11e4-8d92-ba909658ec1e.png)

**Comment by tdev5** [comment]
Or this...no more "upgrading" status...just let "downloaded" be the important thing, but show that it's also snatched.

![status_snatched_downloaded](https://cloud.githubusercontent.com/assets/12114482/7362338/f5a6d6b2-ed56-11e4-90f5-9c03b69ac813.png)

**Comment by TehWardy** [comment]
Daft question:
Why does a "downloaded" row highlight green, theres a column showing that and the quality so no need to highlight yellow either. 

I'm just thinking that maybe the row highlights should be just for things you don't have a column for.
Or even completely removed so the grid can be made to look nicer.

For example:
Wanted: Status "wanted", Downloaded: empty, Snatched: may have quality may not.
Low Qual: Status: Downloaded, Downloaded: quality icon, Snatched: may have quality may not.
Downloaded: Status: Downloaded / Upgrading, Downloaded: quality icon, snatched: may have qual may not.
Skipped: Status: skipped, Downloaded: empty, Snatched: empty
Snatched: Status: Snatched, Downloaded: empty, Snatched: quality snatched icon.
Unaired: as is.

Then row colours could just be a normal "alternating row style grid".
I personally would find this much easier to read.

**Comment by TehWardy** [comment]
That also gives you room for more statuses in the future without having to add to this rainbow of colour. 
It's only suggestion, it comes from me facing this situation at work, we ended up wit ha grid that had about 20 different row colours and it just got ridiculous to look at like it was a childs artwork or something.

**Comment by tdev5** [comment]
>  Why does a "downloaded" row highlight green, theres a column showing that and the quality so no need to highlight yellow either.

You do realize I didn't change this from the way it is currently, right? The only thing I'm doing differently in the last mockup I posted is separating the unified "status" column into separate status, downloaded, and snatched columns.

Actually, I like the color coding, and I think it's pretty good. It helps you visually parse things "at a glance". I could elaborate, if needed.

> I'm just thinking that maybe the row highlights should be just for things you don't have a column for.
> Or even completely removed so the grid can be made to look nicer.

As above, I disagree. 

>  Wanted: Status "wanted", Downloaded: empty, Snatched: may have quality may not.

Either you're suggesting doing away with the "snatched" status, or I don't think this makes sense. But then you list the snatched status later...

For minimal deviation from the current method, if there is an icon in the snatched column, then the status would be snatched. If you think there's a way to improve, please clarify.

As to the rest of your example, I don't see how it differs from my mockup. But it is hard to read and picture what you're saying. Maybe you can make a picture like I did. I used MS Paint.

> Then row colours could just be a normal "alternating row style grid".
> I personally would find this much easier to read.

Again, I disagree. I don't know what other users think. I suspect I'm in the majority. But maybe a different theme would satisfy your desire for less color. I'm not opposed to customization...

> That also gives you room for more statuses in the future without having to add to this rainbow of colour.

I don't think it's an issue yet. Again, my last mockup has no additional statuses, and no additional colors. However, I did purposely give it more variety than you'd normally see, within one show's season, to demonstrate the various possibilities. It would normally look less colorful than my example.

> If you still want to use row colours perhaps use them for "nothing happening", "something happened and all done", "something happening right now". Then its basic traffic light red and amber green style.

I didn't change this. Red green amber blue lavendar. It's almost there now, and I think the current meanings are pretty intuitive.

Still, aren't the colors just a theming issue? Adding the columns isn't, and I think that's the important part of this feature request.

**Comment by TehWardy** [comment]
Ok I will look at mocking something up later as a suggestion. I see your point about not wanting to completely overhaul how things work, I'm just trying to think "in an ideal world, how would I do this ..." then compare that to what's here.
Having faced some of the problems with this type of UI solution before I guess I was just trying to throw that experience at this ticket, it sounds like I didn't explain it very well though. 

I don't think the change needs to be insanely complex or large.

**Comment by tdev5** [comment]
As for the mockup, please feel free to start with mine, and post your suggested changes to that. That will make comparison easier.

**Comment by OmgImAlexis** [comment]
We have switch to using feathub for feature requests. If this feature is still important to anyone, please
check that it has not been added and add it [HERE](http://feathub.com/SiCKRAGETV/SickRage)
Make sure to vote!
